### DISCUSSION TITLE
Is there a way to get matching columns?

### DISCUSSION THREAD

**Discussion by 93Akkord** [moj-analytical-services/splink#1384] [Q&A] [answered]
When running predict, is there a way to get a new column that lists out the columns that matched in the blocking rules?

Example of what I am trying to achieve below.

Left Input
| key | first_name | last_name | age |
| --- | ---------- | --------- | --- |
| 1   | John       | Smith     | 21  |

Right Input
| key | first_name | last_name | age |
| --- | ---------- | --------- | --- |
| 2   | Sam        | Smith     | 21  |

Result
| key_l | key_r | first_name_l | first_name_r | last_name_l | last_name_r | age_l | age_r | matches              |
| ----- | ----- | ------------ | ------------ | ----------- | ----------- | ----- | ----- | -------------------- |
| 1     | 2     | John         | Sam          | Smith       | Smith       | 21    | 21    | ['last_name', 'age'] |

**Comment by RobinL** [comment, marked as answer]
Hiya.

There are various ways to get what you want:

```python
import pandas as pd

from splink.duckdb.linker import DuckDBLinker
from splink.duckdb.duckdb_comparison_library import exact_match


data = [
    {"unique_id": "1", "first_name": "John", "last_name": "Smith", "age": 25},
    {"unique_id": "2", "first_name": "Emily", "last_name": "Johnson", "age": 32},
    {"unique_id": "3", "first_name": "Michael", "last_name": "Williams", "age": 45},
    {"unique_id": "4", "first_name": "John", "last_name": "Smith", "age": 28},
    {"unique_id": "5", "first_name": "Sarah", "last_name": "Davis", "age": 38},
    {"unique_id": "6", "first_name": "John", "last_name": "Smith", "age": 25},
    {"unique_id": "7", "first_name": "Andrew", "last_name": "Johnson", "age": 29},
    {"unique_id": "8", "first_name": "Jessica", "last_name": "Taylor", "age": 32},
    {"unique_id": "9", "first_name": "John", "last_name": "Smith", "age": 25},
    {"unique_id": "10", "first_name": "David", "last_name": "Anderson", "age": 37},
]
df = pd.DataFrame(data)
settings = {
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.first_name = r.first_name and l.last_name = r.last_name",
        "l.age = r.age",
    ],
    "comparisons": [
        exact_match("first_name"),
        exact_match("last_name"),
        exact_match("age"),
    ]
}


linker = DuckDBLinker(df, settings)
print(linker.predict().as_pandas_dataframe().to_markdown(index=False))
```

|   match_weight |   match_probability |   unique_id_l |   unique_id_r | first_name_l   | first_name_r   |   gamma_first_name | last_name_l   | last_name_r   |   gamma_last_name |   age_l |   age_r |   gamma_age |   match_key |
|---------------:|--------------------:|--------------:|--------------:|:---------------|:---------------|-------------------:|:--------------|:--------------|------------------:|--------:|--------:|------------:|------------:|
|       16.7124  |            0.999991 |             1 |             9 | John           | John           |                  1 | Smith         | Smith         |                 1 |      25 |      25 |           1 |           0 |
|        1.71243 |            0.766198 |             4 |             9 | John           | John           |                  1 | Smith         | Smith         |                 1 |      28 |      25 |           0 |           0 |
|       16.7124  |            0.999991 |             6 |             9 | John           | John           |                  1 | Smith         | Smith         |                 1 |      25 |      25 |           1 |           0 |
|       16.7124  |            0.999991 |             1 |             6 | John           | John           |                  1 | Smith         | Smith         |                 1 |      25 |      25 |           1 |           0 |
|        1.71243 |            0.766198 |             4 |             6 | John           | John           |                  1 | Smith         | Smith         |                 1 |      28 |      25 |           0 |           0 |
|        1.71243 |            0.766198 |             1 |             4 | John           | John           |                  1 | Smith         | Smith         |                 1 |      25 |      28 |           0 |           0 |
|      -13.2876  |            0.0001   |             2 |             8 | Emily          | Jessica        |                  0 | Johnson       | Taylor        |                 0 |      32 |      32 |           1 |           1 |

**NOTE:** scroll right on the table above, there are lots more columns you can't see

## Use the comparison vector values:

Note that in the above table, there are columns beginning with gamma_. The values in these columns refer to the match status. For a two-comparison-level column, a value of 1 means 'exact match' and a value of 0 means 'no match'.

Therefore, to some extent, the columns gamma_first_name, gamma_last_name, gamma_age already contain what you want.

If you really need it in the format you specified, you can run additional sql queries on top of the output e.g.:
```
df_predict = linker.predict()
sql = f"""
SELECT
    CASE WHEN gamma_first_name = 1 THEN ARRAY['first_name'] ELSE ARRAY[] END ||
    CASE WHEN gamma_last_name = 1 THEN ARRAY['last_name'] ELSE ARRAY[] END || 
    CASE WHEN gamma_age = 1 THEN ARRAY['age'] ELSE ARRAY[] END  
    AS name_list
FROM
    {df_predict.physical_name}

"""
linker.query_sql(sql)
```

| matches                            |
|:-----------------------------------|
| ['first_name', 'last_name', 'age'] |
| ['first_name', 'last_name']        |
| ['first_name', 'last_name', 'age'] |
| ['first_name', 'last_name', 'age'] |
| ['first_name', 'last_name']        |
| ['first_name', 'last_name']        |
| ['age']                            |

## match_key
Note also that the match_key column (the rightmost column of the output of `predict()`) contains an integer that corresponds to
the (zero based) blocking rules i.e.

match key 0: "l.first_name = r.first_name and l.last_name = r.last_name",
match_key 1: "l.age = r.age","

Note also these are mutually exclusive - i.e. match_key = 1 means the record did not pass match key 0, but did pass match key 1.

**Reply by 93Akkord** [reply]
This helps a lot.  Especially the match_key explanation.  I was thinking about it all wrong.  Thanks!!
